### ISSUE TITLE
False positive for error: TS2873: This kind of expression is always falsy.

### ISSUE THREAD

**Issue by dasa** [microsoft/TypeScript#59505]
### 🔎 Search Terms

"TS2873"

### 🕗 Version & Regression Information

Tested in 5.6.0-beta and 5.6.0-dev.20240801

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.6.0-dev.20240801#code/DYUwLgBAbghsCuIBcEYDsCeEC8E32GAG4AoEgYwHs0BnS0AOmEoHMAKN2BECAH14jw0AExAAzAJZoQwgJQQAZAogBGCAHpocRLKIbNAWgB8gkeKkyK1Oo2bsuiRcrWaHIXfojG8BYCSA

### 💻 Code

```ts
let value: any = null;

console.log((value || undefined) && 1 / value); // -> undefined
console.log(value && 1 / value); // -> null
```

### 🙁 Actual behavior

"TS2873: This kind of expression is always falsy." is reported for `undefined` in the first log call.

### 🙂 Expected behavior

No error since `undefined` is affecting the behavior. Without `undefined` the expression evaluates to `null`.

**Comment by piecyk** [comment]
@RyanCavanaugh still we can get for cases like 

https://www.typescriptlang.org/play/?ts=5.6.2#code/C4TwDgpgBAyglgWzAGwgJQK6qgXigZ2ACc4A7AcygB8oAjAe3tQENTqoNSATCAMzIhd2pLMgBQYgMb1ShKLNxQAPGigQAHsAjd8sRCnRYIAPgAUGyJK1cAXAWJlyAGihEjdtAEpcxqAG8xKCgAemCoABUAeQARSLEAXwlZUwByFJdeZmR8aAAyXKgUm15GOmYiKBsAL3pgFM8xZJSEEAB9EvoyipbWmrqMrJzqGiKOrsq++sb8VPSoYgxoKhHi0tpyidqppp6x9e62yZcFvILRtY3qrYamuc4eflJBKHzC1c79zbqbmea2vY2PSOHG4fAEQmWbwBFSu32msxcImQyBeZ3e41h21+uwuB16W0RomGUNxXymQA

**Comment by RyanCavanaugh** [comment]
@piecyk no plans to change this. The type system isn't sound so there are many places where you might legitimately need to check a variable that the checker thinks is `null` but isn't. The check is only in place to detect syntactic errors, not semantic ones.
